Content.Server/GameObjects/Components/BaseComputerUserInterfaceComponent.cs

```diff
@@ -0,0 +1,126 @@
+using Content.Server.GameObjects.Components.Construction;
+using Content.Server.GameObjects.Components.Power.ApcNetComponents;
+using Content.Server.Utility;
+using Content.Shared.GameObjects.Components;
+using Content.Shared.GameObjects.EntitySystems;
+using Content.Shared.GameObjects.EntitySystems.ActionBlocker;
+using Content.Shared.Interfaces;
+using Content.Shared.Interfaces.GameObjects.Components;
+using Robust.Server.GameObjects;
+using Robust.Shared.Containers;
+using Robust.Shared.GameObjects;
+using Robust.Shared.ViewVariables;
+using Robust.Shared.Localization;
+
+namespace Content.Server.GameObjects.Components
+{
+    /// <summary>
+    /// This component is used as a base class for classes like SolarControlConsoleComponent.
+    /// These components operate the server-side logic for the "primary UI" of a computer.
+    /// That means showing the UI when a user activates it, for example.
+    /// </summary>
+    [ComponentReference(typeof(IActivate))]
+    public abstract class BaseComputerUserInterfaceComponent : Component, IActivate
+    {
+        // { get; private set; } doesn't really express this properly.
+        protected readonly object UserInterfaceKey;
+
+        [ViewVariables] protected BoundUserInterface? UserInterface => Owner.GetUIOrNull(UserInterfaceKey);
+        [ViewVariables] public bool Powered => !Owner.TryGetComponent(out PowerReceiverComponent? receiver) || receiver.Powered;
+
+        public BaseComputerUserInterfaceComponent(object key)
+        {
+            UserInterfaceKey = key;
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            if (UserInterface != null)
+                UserInterface.OnReceiveMessage += OnReceiveUnfilteredUserInterfaceMessage;
+
+            Owner.EnsureComponent<PowerReceiverComponent>();
+        }
+
+        /// <summary>
+        /// Override this to handle messages from the UI before filtering them.
+        /// Calling base is necessary if you want this class to have any meaning.
+        /// </summary>
+        protected void OnReceiveUnfilteredUserInterfaceMessage(ServerBoundUserInterfaceMessage obj)
+        {
+            // "Across all computers" "anti-cheats" ought to be put here or at some parent level (BaseDeviceUserInterfaceComponent?)
+            if (!Powered)
```

Comment:
Can you make this send a popup saying the computer is unpowered?

Reply:
Done.